File: c++/pe020.cpp
// Copyright 2012 Bob Wilkinson <bob@fourtheye.org>

// This is a solution for http://projecteuler.net/problem=20

#include <cstdio>
#include <string>
#include <sstream>
#include <cassert>

#include "./numstring.h"

numstring factorial(const int);

int main() {
  const int maxa = 100;

  numstring fac_string = factorial(maxa);
  int sum = 0;
  for (unsigned int i = 0; i < fac_string().length(); ++i)
    sum += fac_string().at(i) - '0';
  printf("%d\n", sum);

  return(0);
}

File: haskell/pe020.hs
-- this computes projecteuler.net problem 020
import Data.Char

mysum :: String -> Int -> Int
mysum [] n = n
mysum (x:xs) n = mysum xs (digitToInt(x) + n)

factorial :: (Integral a) => a -> a
factorial 0 = 1
factorial n = n * factorial (n - 1)

pe020:: Integer -> Int
pe020 n = mysum (show (factorial n)) 0

main = print(pe020 100)
